### DISCUSSION TITLE
thread/reentrantLock

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#263] [Announcements]
# thread/reentrantLock

重入锁（ReentrantLock）是 Java 并发编程中的一个重要概念。它是一种同步机制，提供了与 synchronized 关键字相同的基本行为，但更灵活，功能也更丰富。

https://javabetter.cn/thread/reentrantLock.html

**Comment by swqddx** [comment]
个人使用jdk23测试， synchonrized的性能应该已经稳定高于RetrantLock了

**Comment by Ricehahaa** [comment]
ReentrantLock是轻量级还是重量级

**Reply by LFAW233** [reply]
轻量级锁跟重量级锁只是synchronized的实现细节，并非实际的锁分类。更好的表述应该是锁自旋，还是阻塞，ReentrantLock基本可以理解为有限自旋+阻塞：
ReentrantLock基于AQS实现，只有当新节点加入后，前驱节点是头节点时才会有限次数地自旋，有限次数内失败则被阻塞。若前驱节点并非头节点，则直接阻塞。说人话就是，如果试图获取锁的线程发现之前没有任何线程在等待锁的释放，那么就会尝试自旋获取锁。
比喻一下，厕所马桶是资源，厕所门就是锁，上厕所的人是希望获取资源的线程。A进入了厕所，锁上了门(A获取锁)，之后B来了，发现厕所里有人(获取锁失败)，他站在门口(成为AQS队列中，挂在头节点后的第一个节点)，敲了几下门试图进入厕所(B有限自旋)，但是里面的人没出来，就停止了敲门(阻塞)。而B之后不知什么时候起，排上了C，C来的时候看到前面已经排着B了(前驱节点不为头节点)，那么就不会去敲B的脑壳让他快一点出来(直接阻塞不自旋)
所以资源占用时间长，且并发高的情况下，阻塞的情况多于自旋的情况。

**Comment by QYJ-795** [comment]
作者你好：
        ReentrantLock的使用这个小节中，对ReentrantLock可重入性代码的验证里，可重入性不是指的是同一个线程可以重复获得锁嘛，代码中是两个线程分别获得锁访问count这个资源保证count的原子性，我似乎没看出来可重入性。

**Reply by LFAW233** [reply]
嗯，感觉应该用递归来测可重入嘛

**Reply by Liucc-123** [reply]
作者这里的示例代码是为了演示ReentrantLock是如何使用的，是否实现了串行操作，并不是对锁的可重入性进行验证。如果想要验证锁的可重入性，需要自己设计测试代码。

**Reply by LFAW233** [reply]
并非如此，在紧跟着代码的“运行结果”一栏之后，作者有明确写到“可以看到，两个线程对 count 变量进行了 20000 次累加操作，说明 ReentrantLock 是支持重入性的。”
由此可见，作者本意是希望验证重入性，而非如你所说，仅仅是演示ReentrantLock如何使用。
